src/parser/mod.rs

```diff
@@ -4139,6 +4160,194 @@ impl<'a> Parser<'a> {
+            include_each,
+            condition,
+            exec_body,
+            characteristics: (deferrable.is_some() || initially.is_some()).then_some(
+                DeferrableCharacteristics {
+                    deferrable,
+                    initially,
+                },
+            ),
+        })
+    }
+
+    pub fn parse_trigger_period(&mut self) -> Result<TriggerPeriod, ParserError> {
+        self.expect_one_of_keywords(&[Keyword::BEFORE, Keyword::AFTER, Keyword::INSTEAD])
+            .and_then(|keyword| {
+                Ok(match keyword {
+                    Keyword::BEFORE => TriggerPeriod::Before,
+                    Keyword::AFTER => TriggerPeriod::After,
+                    Keyword::INSTEAD => self
+                        .expect_keyword(Keyword::OF)
+                        .map(|_| TriggerPeriod::InsteadOf)?,
+                    _ => unreachable!(),
+                })
+            })
+    }
+
+    pub fn parse_trigger_event(&mut self) -> Result<TriggerEvent, ParserError> {
+        self.expect_one_of_keywords(&[
+            Keyword::INSERT,
+            Keyword::UPDATE,
+            Keyword::DELETE,
+            Keyword::TRUNCATE,
+        ])
+        .and_then(|keyword| {
+            Ok(match keyword {
+                Keyword::INSERT => TriggerEvent::Insert,
+                Keyword::UPDATE => {
+                    if self.parse_keyword(Keyword::OF) {
+                        let cols = self.parse_comma_separated(|ident| {
+                            Parser::parse_identifier(ident, false)
+                        })?;
+                        TriggerEvent::Update(cols)
+                    } else {
+                        TriggerEvent::Update(vec![])
+                    }
+                }
+                Keyword::DELETE => TriggerEvent::Delete,
+                Keyword::TRUNCATE => TriggerEvent::Truncate,
+                _ => unreachable!(),
+            })
+        })
+    }
+
+    pub fn parse_trigger_referencing(&mut self) -> Result<Option<TriggerReferencing>, ParserError> {
+        let refer_type = match self.parse_one_of_keywords(&[Keyword::OLD, Keyword::NEW]) {
+            Some(Keyword::OLD) if self.parse_keyword(Keyword::TABLE) => {
+                TriggerReferencingType::OldTable
+            }
+            Some(Keyword::NEW) if self.parse_keyword(Keyword::TABLE) => {
+                TriggerReferencingType::NewTable
+            }
+            _ => {
+                return Ok(None);
+            }
+        };
+
+        let is_as = self.parse_keyword(Keyword::AS);
+        let transition_relation_name = self.parse_object_name(false)?;
+        Ok(Some(TriggerReferencing {
+            refer_type,
+            is_as,
+            transition_relation_name,
+        }))
+    }
+
+    pub fn parse_trigger_exec_body(&mut self) -> Result<TriggerExecBody, ParserError> {
+        let exec_type = self
+            .expect_one_of_keywords(&[Keyword::FUNCTION, Keyword::PROCEDURE])
```

Comment:
I think this would be another candidate to flatten?

Reply:
Done